internal/sink/sink.go

```diff
@@ -0,0 +1,544 @@
+	}
+	if config.MaxCommitWait == 0 {
+		config.MaxCommitWait = 1 * time.Second
+	}
+	if config.NamespaceRefetch == 0 {
+		config.NamespaceRefetch = 15 * time.Second
+	}
+	if config.DeadletterTopicTemplate == "" {
+		config.DeadletterTopicTemplate = defaultDeadletterTopicTemplate
+	}
+
+	sink := &Sink{
+		consumer:       consumer,
+		producer:       producer,
+		config:         config,
+		buffer:         NewSinkBuffer(),
+		namespaceStore: NewNamespaceStore(),
+	}
+
+	return sink, nil
+}
+
+func (s *Sink) flush() error {
+	logger := s.config.Logger.With("sink", "flush")
+	messages := s.buffer.Dequeue()
+
+	// Stop polling new messages from Kafka until we finalize batch
+	s.clearFlushTimer()
+	defer s.setFlushTimer()
+
+	// Dedupe messages so if we have multiple messages in the same batch
+	dedupedMessages := dedupeSinkMessages(messages)
+
+	// 1. Sink to Storage
+	if len(dedupedMessages) > 0 {
+		logger.Debug("started to flush", "buffer size", len(dedupedMessages))
+
+		deadletterMessages := []SinkMessage{}
+		batchesPerNamespace := map[string][]SinkMessage{}
+
+		// Insert per namespace
+		for _, message := range dedupedMessages {
+			if message.Error != nil {
+				switch message.Error.ProcessingControl {
+				case DEADLETTER:
+					logger.Debug("deadlettering message", "error", message.Error, "message", string(message.KafkaMessage.Value), "namespace", message.Namespace)
+					deadletterMessages = append(deadletterMessages, message)
+					continue
+				case DROP:
+					// Do nothing
+					logger.Debug("dropping message", "error", message.Error, "message", string(message.KafkaMessage.Value), "namespace", message.Namespace)
+					continue
+				default:
+					return fmt.Errorf("unknown error type: %s", message.Error)
+				}
+			}
+
+			batchesPerNamespace[message.Namespace] = append(batchesPerNamespace[message.Namespace], message)
+		}
+
+		for namespace, batch := range batchesPerNamespace {
+			list := []*serializer.CloudEventsKafkaPayload{}
+			for _, message := range batch {
+				list = append(list, message.Serialized)
+			}
+
+			err := s.config.Storage.BatchInsert(s.config.Context, namespace, list)
+			if err != nil {
+				// Note: a single error in batch will make the whole batch fail
+				if perr, ok := err.(*ProcessingError); ok {
+					switch perr.ProcessingControl {
+					case DEADLETTER:
+						deadletterMessages = append(deadletterMessages, batch...)
+					case DROP:
+						continue
+					default:
+						return fmt.Errorf("unknown error type: %s", err)
+					}
+				} else {
+					// Throwing and error means we will retry the whole batch again
+					return fmt.Errorf("failed to sink to storage: %s", err)
+				}
+			}
+			logger.Debug("succeeded to sink to storage", "buffer size", len(messages))
+		}
+
+		if len(deadletterMessages) > 0 {
+			err := s.deadLetter(deadletterMessages...)
+			if err != nil {
+				return fmt.Errorf("failed to deadletter messages: %s", err)
+			}
+		}
+	}
+
+	// 2. Commit Offset to Kafka
```

Comment:
I wonder if it would make sense to introduce some sort of retry logic here.

Reply:
It's fine, retry is basically that the batch will be processed again as offset doesn't update.